packages/next/server/image-optimizer.ts

```diff
@@ -48,489 +45,493 @@ try {
-  inflightRequests.set(hash, dedupe.promise)
 
-  try {
-    if (await fileExists(hashDir, 'directory')) {
-      const files = await promises.readdir(hashDir)
-      for (let file of files) {
-        const [maxAgeStr, expireAtSt, etag, extension] = file.split('.')
-        const maxAge = Number(maxAgeStr)
+  async get(cacheKey: string) {
+    try {
+      const cacheDir = join(this.cacheDir, cacheKey)
+      const files = await promises.readdir(cacheDir)
+      const now = Date.now()
+
+      for (const file of files) {
+        const [maxAgeSt, expireAtSt, etag, extension] = file.split('.')
+        const buffer = await promises.readFile(join(cacheDir, file))
         const expireAt = Number(expireAtSt)
-        const contentType = getContentType(extension)
-        const fsPath = join(hashDir, file)
-        xCache = now < expireAt ? 'HIT' : 'STALE'
-        const result = setResponseHeaders(
-          req,
-          res,
-          url,
-          etag,
-          maxAge,
-          contentType,
-          isStatic,
-          isDev,
-          xCache
-        )
-        if (!result.finished) {
-          await new Promise<void>((resolve, reject) => {
-            createReadStream(fsPath)
-              .on('end', resolve)
-              .on('error', reject)
-              .pipe(res)
-          })
-        }
-        if (xCache === 'HIT') {
-          return { finished: true }
-        } else {
-          await promises.unlink(fsPath)
+        const maxAge = Number(maxAgeSt)
+
+        return {
+          value: {
+            kind: 'IMAGE',
+            etag,
+            buffer,
+            extension,
+          },
+          revalidate: maxAge,
+          isStale: now > expireAt,
         }
       }
+    } catch (_) {
+      // failed to read from cache dir, treat as cache miss
     }
+    return null
+  }
+  async set(
+    cacheKey: string,
+    value: {
+      etag: string
+      buffer: Buffer
+      extension: string
+    },
+    revalidate?: number | false
+  ) {
+    if (typeof revalidate !== 'number') {
+      throw new Error('invariant revalidate must be a number for image-cache')
+    }
+    const expireAt =
+      Math.max(revalidate, this.nextConfig.images.minimumCacheTTL) * 1000 +
+      Date.now()
+
+    await writeToCacheDir(
+      join(this.cacheDir, cacheKey),
+      value.extension,
+      revalidate,
+      expireAt,
+      value.buffer,
+      value.etag
+    )
+  }
+}
+export class ImageError extends Error {
+  statusCode: number
 
-    let upstreamBuffer: Buffer
-    let upstreamType: string | null
-    let maxAge: number
-
-    if (isAbsolute) {
-      const upstreamRes = await fetch(href)
-
-      if (!upstreamRes.ok) {
-        res.statusCode = upstreamRes.status
-        res.end('"url" parameter is valid but upstream response is invalid')
-        return { finished: true }
-      }
+  constructor(message: string, statusCode: number) {
+    super(message)
+    this.statusCode = statusCode
```

Comment:
Minor nit pick: can we swap the params so that statusCode is first? I think it will read better.

Reply:
Updated in https://github.com/vercel/next.js/pull/34075/commits/cfb06f1e79ff38d23480adbe198d1fce4d7733ce